assets/src/hooks/useApiCall.ts

```diff
@@ -0,0 +1,65 @@
+import { useEffect, useState } from "react"
+
+/**
+ * Return type of {@linkcode useApiCall}
+ */
+interface ApiLoadingResult<T> {
+  /**
+   * Indicates if {@linkcode useApiCall} is in the process of computing a new
+   * value
+   */
+  isLoading: boolean
+
+  /**
+   * The last value that {@linkcode useApiCall} resolved.
+   *
+   * Is `undefined` when it has not resolved or it resolves to `undefined`
+   */
+  result: T | undefined
+}
+
+/** Hook prop arguments for {@linkcode useApiCall} */
+interface UseApiCallProps<T> {
+  /**
+   * Function which returns a promise that is called every time it changes.
+   * @param abortSignal An {@linkcode AbortSignal} that signals if the promise
+   *    should attempt to cancel early
+   */
+  apiCall: (abortSignal: AbortSignal) => Promise<undefined | T>
+}
+
+/**
+ * Calls the provided function returning a {@linkcode Promise} and tracks the
+ * loading state and the previously resolved value.
+ *
+ * @returns {ApiLoadingResult<T>}
+ */
+export const useApiCall = <T>({
+  apiCall,
+}: UseApiCallProps<T>): ApiLoadingResult<T> => {
```

Comment:
I'm actually really conflicted on merging this now. We've needed something like this for a while, but if we can wait a little longer, once `use` gets out of the canary channel, we'll have a much better way of handling promises and loading states via react's `Suspense` features.
If I could figure out a way to polyfill `use` or make this interact with `Suspense` well I'd be more inclined to merge it, but I now am not sure that this is worth it, because I would prefer to rip out any usages of this afterward for `use` once it's available

https://react.dev/reference/react/use
https://react.dev/reference/react/useDeferredValue
https://react.dev/reference/react/Suspense

Reply:
How long does it usually take for features to get through the canary channel?

Reply:
I don't know how long it takes, but it does look like it's encouraged to use canary if you want to be ready for new features when they land or try them out early. It's more stable than experimental branches and bugs are treated the same as release channels. So maybe that's a good reason to try the canary release?

Reply:
https://react.dev/blog/2023/05/03/react-canaries

Reply:
[for posterity]: we decided we're uncomfortable with depending on a canary channel in an ops tool at this moment.